Readme.md
# Bezierpp

Header only library to create bezier curve, in 2D or 3D.

## Usage

> Examples can be found in `examples` folder. Integration usage is shown with [Eigen](https://eigen.tuxfamily.org) and [Qt](https://www.qt.io/).

### Lerp & InvLerp

The library provide convenient lerp and invLerp functions.

```cpp
bezierpp::lerp( a, b, t ) = value
bezierpp::invLerp( a, b, value ) = t
```

![](https://uploads.gamedev.net/monthly_2019_10/1.gif.4cd2689cefca32175d9b58fca4f45fe7.gif)

### Linear Curve

```cpp
#include <bezierpp/bezierpp.hpp>
#include <Eigen/Core>
#include <iostream>

constexpr int segmentCount = 10;
const Eigen::Vector2d p1(1, 0);
const Eigen::Vector2d p2(0, 1);

for(int i = 0; i < segmentCount; ++i)
{
  const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
  const auto p = bezierpp::linearBezierCurve(p1, p2, t);
  std::cout << p << std::endl;
}
```

![Animation of a linear Bézier curve, *t* in [0,1]](https://upload.wikimedia.org/wikipedia/commons/thumb/0/00/B%C3%A9zier_1_big.gif/240px-B%C3%A9zier_1_big.gif)

### Quadratic Curve

```cpp
#include <bezierpp/bezierpp.hpp>
#include <Eigen/Core>
#include <iostream>

constexpr int segmentCount = 10;
const Eigen::Vector2d p1(0, 0);
const Eigen::Vector2d p2(0.4, 1);
const Eigen::Vector2d p3(1, 0);
for(int i = 0; i < segmentCount; ++i)
{
    const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
    const auto p = bezierpp::quadraticBezierCurve(p1, p2, p3, t);
    std::cout << i << " : " << p.format(CleanFmt) << std::endl;
}
```



![Animation of a quadratic Bézier curve, *t* in [0,1]](https://upload.wikimedia.org/wikipedia/commons/thumb/3/3d/B%C3%A9zier_2_big.gif/240px-B%C3%A9zier_2_big.gif)

### Cubic Curve

```cpp
#include <bezierpp/bezierpp.hpp>
#include <Eigen/Core>
#include <iostream>

const Eigen::Vector2d p1(0, 0);
const Eigen::Vector2d p2(-0.1, 1);
const Eigen::Vector2d p3(0.8, 1);
const Eigen::Vector2d p4(1, 0);
for(int i = 0; i < segmentCount; ++i)
{
    const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
    const auto p = bezierpp::cubicBezierCurve(p1, p2, p3, p4, t);
    std::cout << i << " : " << p.format(CleanFmt) << std::endl;
}
```



![Animation of a cubic Bézier curve, *t* in [0,1]](https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/B%C3%A9zier_3_big.gif/240px-B%C3%A9zier_3_big.gif)

### Smooth Spline

If you want to draw smooth spline as describe in the [article](http://scaledinnovation.com/analytics/splines/aboutSplines.html), a convenient `getBezierControlPoints` is provided by the library.

![](http://scaledinnovation.com/analytics/splines/splineDefault.png)

The `getBezierControlPoints` require the norm of the vector, so `getNorm` should be specialized if using type different than Eigen. For example with Qt QVector3D:

```cpp
#include <QtGui/QVector3D>

namespace bezierpp{
template<>
float getNorm<QVector3D, float>(const QVector3D& value)
{
    return value.length();
}
}
```

For convenience, specialization is already provided for qt in `bezierpp/qt.hpp`

### Examples

Eigen example output:

```
Linear Curve :
0 : 1, 0
1 : 0.888889, 0.111111
2 : 0.777778, 0.222222
3 : 0.666667, 0.333333
4 : 0.555556, 0.444444
5 : 0.444444, 0.555556
6 : 0.333333, 0.666667
7 : 0.222222, 0.777778
8 : 0.111111, 0.888889
9 : 0, 1

Quadratic Curve :
0 : 0, 0
1 : 0.091358, 0.197531
2 : 0.187654, 0.345679
3 : 0.288889, 0.444444
4 : 0.395062, 0.493827
5 : 0.506173, 0.493827
6 : 0.622222, 0.444444
7 :  0.74321, 0.345679
8 : 0.869136, 0.197531
9 : 1, 0

Cubic Curve :
0 : 0, 0
1 : 0.00137174, 0.296296
2 : 0.0628258, 0.518519
3 : 0.17037, 0.666667
4 : 0.310014, 0.740741
5 : 0.467764, 0.740741
6 : 0.62963, 0.666667
7 : 0.781619, 0.518519
8 : 0.909739, 0.296296
9 : 1, 0
```

Qt example output:

```
Linear Curve :
0  :  QVector2D(1, 0)
1  :  QVector2D(0.888889, 0.111111)
2  :  QVector2D(0.777778, 0.222222)
3  :  QVector2D(0.666667, 0.333333)
4  :  QVector2D(0.555556, 0.444444)
5  :  QVector2D(0.444444, 0.555556)
6  :  QVector2D(0.333333, 0.666667)
7  :  QVector2D(0.222222, 0.777778)
8  :  QVector2D(0.111111, 0.888889)
9  :  QVector2D(0, 1)

Quadratic Curve :
0  :  QVector2D(0, 0)
1  :  QVector2D(0.091358, 0.197531)
2  :  QVector2D(0.187654, 0.345679)
3  :  QVector2D(0.288889, 0.444444)
4  :  QVector2D(0.395062, 0.493827)
5  :  QVector2D(0.506173, 0.493827)
6  :  QVector2D(0.622222, 0.444444)
7  :  QVector2D(0.74321, 0.345679)
8  :  QVector2D(0.869136, 0.197531)
9  :  QVector2D(1, 0)

Cubic Curve :
0  :  QVector2D(0, 0)
1  :  QVector2D(0.00137174, 0.296296)
2  :  QVector2D(0.0628258, 0.518519)
3  :  QVector2D(0.17037, 0.666667)
4  :  QVector2D(0.310014, 0.740741)
5  :  QVector2D(0.467764, 0.740741)
6  :  QVector2D(0.62963, 0.666667)
7  :  QVector2D(0.781619, 0.518519)
8  :  QVector2D(0.909739, 0.296296)
9  :  QVector2D(1, 0)
```



## 🔨 How to use

### CPM

The library support [CPM](https://github.com/cpm-cmake/CPM.cmake).

```cmake
include(CPM.cmake)
CPMAddPackage(
  GIT_REPOSITORY https://github.com/OlivierLDff/bezierpp
  GIT_TAG main
)
```

### Header only

Just copy header located in `src` in your project include directory.

### Installation

```bash
git clone https://github.com/OlivierLDff/bezierpp
cmake -E make_directory bezierpp/build
cmake -DCMAKE_PREFIX_PATH=install  \
      -DBEZIERPP_ENABLE_INSTALL=ON \
      -B bezierpp/build -S bezierpp
```

This will create under `install` folder following hierarchy:

```
include
 └─bezierpp
    └─bezierpp.hpp
lib
 └─cmake
    └─bezierpp-version
       ├─bezierppConfig.cmake
       ├─bezierppConfigVersion.cmake
       └─bezierppTargets.cmake
```

## Bibliography

* [Bezier Curve on Wikipedia](https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Constructing_B%C3%A9zier_curves)
* [About Splines](http://scaledinnovation.com/analytics/splines/aboutSplines.html)

## 👥 Author

* [Olivier Le Doeuff](olivier.ldff@gmail.com)

CHANGELOG.md
# Changelog

<a name="1.0.3"></a>
## 1.0.3 (2022-09-28)

### Changed

- 🎨 cmake-format [[d4eb7d4](https://github.com/OlivierLDff/bezierpp/commit/d4eb7d462a708594a597dfd645423157fa37e742)]

### Miscellaneous

- 🔨 Install in CMAKE_INSTALL_INCLUDEDIR to follow modern cmake convention [[cd845cf](https://github.com/OlivierLDff/bezierpp/commit/cd845cf334aa2a8aab724e6aa8f13bfe2e19975a)]
-  👷 Generate changelog with gitmoji [[7a98153](https://github.com/OlivierLDff/bezierpp/commit/7a98153ef6e5f3a1e160273723ceaa61ed564352)]


<a name="1.0.2"></a>
## 1.0.2 (2022-05-19)

### Miscellaneous

- 🔨 Fix build for old cmake version CMakeLists.txt:76 (target_sources):     target_sources may only set INTERFACE properties on INTERFACE targets [[a9a1dc1](https://github.com/OlivierLDff/bezierpp/commit/a9a1dc1f34d1a75924f8fc1bf07f66c32aa63c1d)]


<a name="1.0.1"></a>
## 1.0.1 (2022-04-26)

### Fixed

- 🐛 add missing inline for template specialization [[a1678a1](https://github.com/OlivierLDff/bezierpp/commit/a1678a1e25d98da0b3a798c15507455b0f7b70f7)]


<a name="1.0.0"></a>
## 1.0.0 (2022-04-26)

### Miscellaneous

-  👷 CI [[2dbbfb7](https://github.com/OlivierLDff/bezierpp/commit/2dbbfb7625f650e5343e0e9070a3365d604bc177)]
- 📝 set language to cmake [[22a50d3](https://github.com/OlivierLDff/bezierpp/commit/22a50d3808fb8fa4c5990c988652ab23750e9446)]
- 🌱 Initial Commit [[b2437ff](https://github.com/OlivierLDff/bezierpp/commit/b2437ff97e37dc5c4c20f8bdeb40823a8886c8bb)]




src/bezierpp/qt.hpp
// Copyright 2021 Olivier Le Doeuff
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files
// (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

#ifndef __BEZIERPP_QT_HPP__
#define __BEZIERPP_QT_HPP__

#include <bezierpp/bezierpp.hpp>
#include <QtGui/QVector2D>
#include <QtGui/QVector3D>
#include <QtGui/QVector4D>

namespace bezierpp {

template<>
inline float getNorm<QVector2D, float>(const QVector2D& value)
{
    return value.length();
}

template<>
inline float getNorm<QVector3D, float>(const QVector3D& value)
{
    return value.length();
}

template<>
inline float getNorm<QVector4D, float>(const QVector4D& value)
{
    return value.length();
}

}

#endif


samples/bezierpp_qt.cpp
// Copyright 2021 Olivier Le Doeuff
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files
// (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

#include <bezierpp/bezierpp.hpp>
#include <bezierpp/qt.hpp>
#include <QtCore/QDebug>

int main(int argc, char* argv[])
{
    constexpr int segmentCount = 10;

    // Linear Curve
    {
        qDebug() << "Linear Curve :";

        const QVector2D p1(1, 0);
        const QVector2D p2(0, 1);

        for(int i = 0; i < segmentCount; ++i)
        {
            const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
            const auto p = bezierpp::linearBezierCurve(p1, p2, t);

            qDebug() << i << " : " << p;
        }
    }

    qDebug() << "";

    // Quadratic Curve
    {
        qDebug() << "Quadratic Curve :";

        const QVector2D p1(0, 0);
        const QVector2D p2(0.4, 1);
        const QVector2D p3(1, 0);

        for(int i = 0; i < segmentCount; ++i)
        {
            const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
            const auto p = bezierpp::quadraticBezierCurve(p1, p2, p3, t);

            qDebug() << i << " : " << p;
        }
    }

    qDebug() << "";

    // Cubic Curve
    {
        qDebug() << "Cubic Curve :";

        const QVector2D p1(0, 0);
        const QVector2D p2(-0.1, 1);
        const QVector2D p3(0.8, 1);
        const QVector2D p4(1, 0);

        for(int i = 0; i < segmentCount; ++i)
        {
            const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
            const auto p = bezierpp::cubicBezierCurve(p1, p2, p3, p4, t);

            qDebug() << i << " : " << p;
        }
    }
}

src/bezierpp/bezierpp.hpp
// Copyright 2021 Olivier Le Doeuff
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files
// (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

#ifndef __BEZIERPP_HPP__
#define __BEZIERPP_HPP__

#include <utility>
#include <limits>

namespace bezierpp {

template<typename Scalar>
Scalar invLerp(Scalar a, Scalar b, Scalar v)
{
    if(b == a)
        return {};

    return (v - a) / (b - a);
}

template<typename Scalar, typename ScalarT>
Scalar lerp(const Scalar& a, const Scalar& b, ScalarT t)
{
    return a + t * (b - a);
}

template<typename T, typename Scalar>
Scalar getNorm(const T& value)
{
    return value.norm();
}

// \brief Get optimal bezier control points, for smooth bezier curve around p2, with a curve starting at p1 and ending at p3
// More info http://scaledinnovation.com/analytics/splines/aboutSplines.html
// \tparam T Vector data holder that must support  + - * operator
// \tparam Scalar scalar type (float or double)
// \param p1 first point where spline starts
// \param p2 middle point where spline should go
// \param p1 last point where spline ends
// \param tension Control how smooth spline is.
// 0 is linear interpolation between points
// 0.5 is the smoothest as possible
// 1 act like a "radius"
// \return Bezier control points that should be used around p2
template<typename T, typename Scalar>
std::pair<T, T> getBezierControlPoints(const T& p1, const T& p2, const T& p3, float tension = 0.5)
{
    const auto d12 = getNorm<T, Scalar>(p1 - p2);
    const auto d23 = getNorm<T, Scalar>(p2 - p3);
    const auto d13 = d12 + d23;

    if(d13 >= -std::numeric_limits<Scalar>::epsilon() && d13 <= std::numeric_limits<Scalar>::epsilon())
        return {p2, p2};

    const auto fa = tension * d12 / d13;
    const auto fb = tension * d23 / d13;

    const auto p31 = p3 - p1;
    const auto pc1 = p2 - fa * p31;
    const auto pc2 = p2 + fb * p31;

    return {pc1, pc2};
}

// \brief Give position on 3D linear bezier curve following equation:
// P = P1+t(P2-P1)
// See more about bezier curve on wikipedia : https://en.wikipedia.org/wiki/Bézier_curve#Constructing_Bézier_curves
// \tparam T Vector data holder that must support  + - * operator
// \tparam Scalar scalar type (float or double)
// \param p1 is origin point
// \param p2 is destination point
// \param t is expected to be between 0 and 1
// \return Position on point at t
template<typename T, typename Scalar>
T linearBezierCurve(const T& p1, const T& p2, Scalar t)
{
    return lerp<T, Scalar>(p1, p2, t);
}

// \brief Give position on 3D quadratic bezier curve following equation:
// P = (1−t)²P1 + 2(1−t)tP2 + t²P3
// See more about bezier curve on wikipedia : https://en.wikipedia.org/wiki/Bézier_curve#Constructing_Bézier_curves
// \tparam T Vector data holder that must support  + - * operator
// \tparam Scalar scalar type (float or double)
// \param p1 is origin point
// \param p2 is control point
// \param p3 is destination point
// \param t is expected to be between 0 and 1
// \return Position on point at t
template<typename T, typename Scalar>
T quadraticBezierCurve(const T& p1, const T& p2, const T& p3, Scalar t)
{
    // P = (1−t)²P1 + 2(1−t)tP2 + t²P3
    // p =    a     +     b     +  c

    const auto reverseT = (1 - t);

    const auto a = p1 * reverseT * reverseT;
    const auto b = 2 * reverseT * t * p2;
    const auto c = t * t * p3;

    return a + b + c;
}

// \brief Give position on 3D cubic bezier curve following equation:
// P = (1−t)³P1 + 3(1−t)²tP2 + 3(1−t)t²P3 + t³P4
// See more about bezier curve on wikipedia : https://en.wikipedia.org/wiki/Bézier_curve#Constructing_Bézier_curves
// \tparam T Vector data holder that must support  + - * operator
// \tparam Scalar scalar type (float or double)
// \param p1 is origin point
// \param p2 is first control point
// \param p3 is second control point
// \param p4 is destination point
// \param t is expected to be between 0 and 1
// \return Position on point at t
template<typename T, typename Scalar>
T cubicBezierCurve(const T& p1, const T& p2, const T& p3, const T& p4, Scalar t)
{
    // P = (1−t)³P1 + 3(1−t)²tP2 + 3(1−t)t²P3 + t³P4
    // p =     a    +      b     +      c     +   d

    const auto reverseT = (1 - t);

    const auto a = reverseT * reverseT * reverseT * p1;
    const auto b = 3 * reverseT * reverseT * t * p2;
    const auto c = 3 * reverseT * t * t * p3;
    const auto d = t * t * t * p4;

    return a + b + c + d;
}

}

#endif


samples/bezierpp_eigen.cpp
// Copyright 2021 Olivier Le Doeuff
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files
// (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge,
// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

#include <bezierpp/bezierpp.hpp>
#include <Eigen/Core>
#include <iostream>

const Eigen::IOFormat CleanFmt(Eigen::StreamPrecision, 0, ", ", ", ", "", "");

int main(int argc, char* argv[])
{
    constexpr int segmentCount = 10;

    // Linear Curve
    {
        std::cout << "Linear Curve :" << std::endl;

        const Eigen::Vector2d p1(1, 0);
        const Eigen::Vector2d p2(0, 1);

        for(int i = 0; i < segmentCount; ++i)
        {
            const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
            const auto p = bezierpp::linearBezierCurve(p1, p2, t);

            std::cout << i << " : " << p.format(CleanFmt) << std::endl;
        }
    }

    std::cout << std::endl;

    // Quadratic Curve
    {
        std::cout << "Quadratic Curve :" << std::endl;

        const Eigen::Vector2d p1(0, 0);
        const Eigen::Vector2d p2(0.4, 1);
        const Eigen::Vector2d p3(1, 0);

        for(int i = 0; i < segmentCount; ++i)
        {
            const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
            const auto p = bezierpp::quadraticBezierCurve(p1, p2, p3, t);

            std::cout << i << " : " << p.format(CleanFmt) << std::endl;
        }
    }

    std::cout << std::endl;

    // Cubic Curve
    {
        std::cout << "Cubic Curve :" << std::endl;

        const Eigen::Vector2d p1(0, 0);
        const Eigen::Vector2d p2(-0.1, 1);
        const Eigen::Vector2d p3(0.8, 1);
        const Eigen::Vector2d p4(1, 0);

        for(int i = 0; i < segmentCount; ++i)
        {
            const auto t = bezierpp::invLerp<float>(0.f, float(segmentCount - 1), float(i));
            const auto p = bezierpp::cubicBezierCurve(p1, p2, p3, p4, t);

            std::cout << i << " : " << p.format(CleanFmt) << std::endl;
        }
    }
}